### PR TITLE
S3 util improve

### PR DESCRIPTION
**Pull request by mradwan**
Two changes in this branch
- Add logging for different errors encountered when using listObjects
- Add support for using ListObjectsV2 instead and gate it behind gflag

### CODE REVIEW

**Comment by mradwan** [comment]
Test plan:
For testing the additional logging I deployed private build on a cluster and observed the stats as I triggered some errors

Screen Shot 2022-03-25 at 10 55 31 AM

For testing the s3 util change I gated it behind flag and enabled it on test cluster, will observe how it goes.
I also did a local test to ensure continuation works same way by enforcing 100 object limit on the request
```
#include "third_party/rocksplicator/rocksplicator/common/s3util.h"
#include <gflags/gflags.h>
#include <iostream>


int main(int argc, char** argv){
    gflags::ParseCommandLineFlags(&argc, &argv, true);
    auto util = common::S3Util::BuildS3Util(50, "pinterest-jackson");
    auto resp = util->listAllObjects("meariby_test/rockstore_data/1/");
    std::cout << resp.Body().objects.size() << std::endl;
    return 0;
}
```

```
root@dev-meariby:/home/meariby/code/cosmos# ./bazel-bin/sandbox/test_main --use_list_objects_v2
Using list objects v2 (new)
Read 100 objects
Using list objects v2 (new)
Read 100 objects
Using list objects v2 (new)
Read 100 objects
Using list objects v2 (new)
Read 100 objects
Using list objects v2 (new)
Read 100 objects
Using list objects v2 (new)
Read 30 objects
530
root@dev-meariby:/home/meariby/code/cosmos# ./bazel-bin/sandbox/test_main
Using List Objects old
Read 100 objects
Using List Objects old
Read 100 objects
Using List Objects old
Read 100 objects
Using List Objects old
Read 100 objects
Using List Objects old
Read 100 objects
Using List Objects old
Read 30 objects
530
```

**Comment by mradwan** [comment]
Note: I did consider doing a refactor to avoid duplicating the entire function but given that this code is only needed temporarily I found it not worth investing in. but rather focus on rolling out and just deleting the old function.

File: common/s3util.cpp [hunk 1]
```diff
@@ -254,6 +310,10 @@ void S3Util::listObjectsHelper(const string& prefix, const string& delimiter,
     }
   } else {
     if (error_message != nullptr) {
+      Stats::get()->Incr(folly::sformat("s3_list_objects_helper_error response_code={} exception_name={} should_retry={}", 
+        static_cast<int32_t>(listObjectResult.GetError().GetResponseCode()),
+        listObjectResult.GetError().GetExceptionName(), 
+        listObjectResult.GetError().ShouldRetry()));
```
Comment (kangnanli) [inline common/s3util.cpp:316]: Do you have a sample error log after this change?

**Review by kangnanli — APPROVED** [review]
lgtm. @jaricftw, can you also take a look in case I missed anything. 

Fix the cmt before land.

File: common/s3util.cpp [hunk 2]
```diff
@@ -214,6 +217,59 @@ SdkGetObjectResponse S3Util::sdkGetObject(const string& key,
   return getObjectResult;
 }
 
+
+void S3Util::listObjectsV2Helper(const string& prefix, const string& delimiter,
```
Comment (jaricftw) [inline common/s3util.cpp:221]: `listObjectsV2Helper` looks like very much like `listObjectsHelper`, is there anything we can do to DRY this up a bit to reuse common logic?

**Comment by jaricftw** [comment]
> Note: I did consider doing a refactor to avoid duplicating the entire function but given that this code is only needed temporarily I found it not worth investing in. but rather focus on rolling out and just deleting the old function.

oh didn't see this when I made the comment. Yeah, I think this is fair, in this case: can we add a comment on ListObjectHelper to indicate this is being deprecated and will be deleted once we move to ListObjectV2?

**Review by jaricftw — APPROVED** [review]

File: common/s3util.cpp (hunk 1 above)
Comment (mradwan) [reply common/s3util.cpp:316]: @kangnanli this was in my previous pull request https://github.com/pinterest/rocksplicator/pull/602

Comment (mradwan) [reply common/s3util.cpp:316]: Screen Shot 2022-03-25 at 10 55 31 AM

File: common/s3util.cpp (hunk 2 above)
Comment (mradwan) [reply common/s3util.cpp:221]: To be honest the refactor is quite doable with templatizing the request object and the actual list objects call but  but I prefer keeping old function as is and just doing switch over.

**Pull request merged**